cpp/include/raft/matrix/shift.cuh

```diff
@@ -0,0 +1,146 @@
+ * @param[in] values: value matrix of size (n_rows x k) for shift_type=ShiftType::COL or (k x
+ * n_cols) for shift_type=ShiftType::ROW to fill in empty space of in_out after shifting.
+ * @param[in] shift_direction: ShiftDirection::TOWARDS_BEGINNING shifts towards the 0th row/col
+ * direction, and ShiftDirection::TOWARDS_END shifts towards the (nrow-1)th row/col direction
+ * @param[in] shift_type: ShiftType::ROW shifts rows and ShiftType::COL shift columns
+ */
+template <typename ValueT, typename IdxT>
+void shift(raft::resources const& handle,
+           raft::device_matrix_view<ValueT, IdxT, row_major> in_out,
+           raft::device_matrix_view<const ValueT, IdxT> values,
+           ShiftDirection shift_direction = ShiftDirection::TOWARDS_END,
+           ShiftType shift_type           = ShiftType::COL)
+{
+  if (shift_type == ShiftType::COL) {
+    RAFT_EXPECTS(in_out.extent(0) == values.extent(0),
+                 "in_out matrix and the values matrix should haver the same number of rows when "
+                 "using shift_type=ShiftType::COL");
+    RAFT_EXPECTS(in_out.extent(1) > values.extent(1),
+                 "number of columns in in_out should be > number of columns in values when using "
+                 "shift_type=ShiftType::COL");
+  } else {
+    RAFT_EXPECTS(in_out.extent(1) == values.extent(1),
+                 "in_out matrix and the values matrix should haver the same number of cols when "
+                 "using shift_type=ShiftType::ROW");
+    RAFT_EXPECTS(in_out.extent(0) > values.extent(0),
+                 "number of rows in in_out should be > number of rows in values when using "
+                 "shift_type=ShiftType::ROW");
+  }
+
+  detail::shift(handle, in_out, values, shift_direction, shift_type);
+}
+
+/**
+ * @brief In-place operation. Shifts all rows or columns to shift_direction by k, and fills the
+ * first k row or columns with its column id or row id, respectively.
+ *
+ * Example 1) if we have a row-major 3x4 matrix in_out = [[1,2,3,4], [5,6,7,8],
+ * [9,10,11,12]], k=2, shift_direction = ShiftDirection::TOWARDS_END and shift_type =
+ * ShiftType::COL, then we end up with [[0,0,1,2], [1,1,5,6], [2,2,9,10]].
+ * Example 2) if we have a row-major 3x4 matrix in_out = [[1,2,3,4], [5,6,7,8], [9,10,11,12]], k=2,
+ * shift_direction = ShiftDirection::TOWARDS_BEGINNING and shift_type = ShiftType::ROW, then we end
+ * up with [[9,10,11,12], [0,1,2,3], [0,1,2,3]]
+ *
+ * @param[in] handle: raft handle
+ * @param[in out] in_out: input matrix of size (n_rows, n_cols)
+ * @param[in] k: shift size
+ * @param[in] shift_direction: ShiftDirection::TOWARDS_BEGINNING shifts towards the 0th row/col
+ * direction, and ShiftDirection::TOWARDS_END shifts towards the (nrow-1)th row/col direction
+ * @param[in] shift_type: ShiftType::ROW shifts rows and ShiftType::COL shift columns
+ */
+template <typename math_t, typename matrix_idx_t>
+void shift_self(raft::resources const& handle,
```

Comment:
Can we incorporate this as a specialized input into the prior functions so that we can avoid having a separate function just for this? Every template function we expose becomes another compiled kernel into the user's binary. The more we can consolidate, the better.

Reply:
val is not a nullopt. If not specified, we fill with row ids and column ids.